efiction/chapters.py

```diff
@@ -59,25 +60,73 @@ def __load_chapter_text_into_db(self, chapter_paths: List[dict]):
+
+    https://github.com/otwcode/open-doors-mojibake
+
+follow the instructions in readme and then run it with this command:
+
+    mojibake {self.config['Archive']['chapter_path']}
+
+    """.strip()
+)
+            print(message_string)
+            while encoding is None:
+                encoding_text = input("Enter a valid encoding (press enter for utf8): ")
+                if encoding_text == "":
+                    encoding_text = 'utf8'
+                try:
+                    # check if encoding is valid
+                    ''.encode(encoding_text)
+                    encoding = encoding_text
+                except:
+                    print(f"{encoding_text} is not a valid encoding, try again")
         for old_chapter in old_chapters:
             chapid = old_chapter['chapid']
             chapter = [chapter_path for chapter_path in chapter_paths if chapter_path['chap_id'] == str(chapid)]
             if chapter:
                 file = chapter[0]['path']
-                try:
-                    with open(file, 'r', encoding="utf-8") as f:
-                        raw = f.read()
-                except UnicodeDecodeError as err:
-                    warnings.append(f"Chapter with id {chapid} contains non-ASCII characters which are not valid "
-                                    f"UTF-8. Trying Windows 1252...")
-                    try:
-                        with open(file, 'r', encoding='cp1252') as f:
-                            raw = f.read()
-                    except UnicodeDecodeError as err:
-                        warnings.append(f"Chapter with id {chapid} contains non-ASCII characters which are not valid "
-                                        f"Windows 1252. Trying Latin 1...")
-                        with open(file, 'r', encoding='latin-1') as f:
-                            raw = f.read()
+                with open(file, 'rb') as raw_chapter:
+                    raw = raw_chapter.read()
+                    while isinstance(raw, bytes):
+                        try:
+                            raw = raw.decode(encoding=encoding)
+                        except UnicodeDecodeError as e:
+                            error = f"Failed to decode {file}\n"
+                            line_num = raw[:e.start].decode(encoding).count("\n")
+                            error += f"At line {line_num}:\t{str(e)}\n"
+                            error += "--\t" + str(raw[max(e.start - 40, 0):e.end + 30]) + "\n"
+                            # print `^` under the offending byte
+                            error += "\t" + \
+                                     " " * (len(str(raw[max(e.start - 40, 0):e.start])) - 1) + \
+                                     "^" * (len(str(raw[e.start:e.end])) - 3) + "\n"
+                            error += "Will be converted to:\n"
+                            # remove the offending bytes (usually one)
+                            raw = raw[:e.start] + raw[e.end:]
```

Comment:
I moved this warning to the end section, because it is already logging around 2MiB/s of errors via ssh, so this additional message would have significant performance implications.

And It would probably not help much if we were to spam the user with it `210636` times.

Reply:
Yeah, upon reflection, it would probably be most useful very early in the loop, to give the user the option to stop the process and fix the encoding before the other 210636 characters are needlessly processed. After all, if you have more than, say, a dozen invalid characters, you've probably got an encoding problem you must fix!

Might even be nice to prompt the user after the first X messages to point out that there seem to be a lot of invalid characters and ask them if they're sure they want to continue, or would they prefer to go and use mojibake to find out what the encoding is actually likely to be? (with X some lowish number like 3 or 5, for example?)

Reply:
Any warning early in a loop would get probably swept aside by the flood of logs incoming, but abandoning the processing after, let's say `.03 * numberOfStories` errors would be sufficient I think? That would be around a hundred encoding errors for this archive I think.

Reply:
I meant an actual prompt early in the loop, or after `.03 * numberOfStories`. It could say something like "Automatic modifications have been made to {warnings} chapters so far. This may indicate an incorrect encoding: please review the messages above to see the modifications. Are you sure you want to continue (y/n)?". 

If the user says `y`, then carry on with the loop. If they say `n`, maybe print a message asking them to use `mojibake` to find the correct encoding and set it in the .ini file. What do you think?

Reply:
I could do that!